### ISSUE TITLE
How to extend a typedef parameter in JSDOC?

### ISSUE THREAD

**Issue by julmot** [jsdoc3/jsdoc#1199]
Description: http://stackoverflow.com/questions/36737921/how-to-extend-a-typedef-parameter-in-jsdoc

**Comment by tenorok** [comment]
Anybody knows correct way to extend typedefs?

**Comment by katoquro** [comment]
looks like `@augments` (`@extends`) doesn't work

**Comment by dietergeerts** [comment]
Wanted to know this myself, and tried the follwing, which works for IntelliJ intellisense:
```
/**
 * @typedef {Object} Store
 * @property {function(): Rx.Observable.<OBJ[]>} getAll
 * @property {function(DATA): Rx.Observable.<OBJ>} create
 * @template OBJ, DATA
 */

/**
 * @typedef {Store} MeetingStore
 * @property {function(DATA): Rx.Observable.<OBJ>} update
 * @template OBJ, DATA
 */
```

**Comment by MatthewHerbst** [comment]
Has anyone found a solution for this, or is it still not supported?

@dietergeerts did you have to do anything special to get that working for IntelliJ intellisense? I implemented the same pattern but it doesn't work for VS Code intellisense unfortunately - it just shows the original base type, not the base type with the new field(s) merged in

**Comment by FireController1847** [comment]
Yes, I agree. This is something I really would like as it'd make my life 9000% easier.

**Comment by dietergeerts** [comment]
@MatthewHerbst We also have devs in our team that use VS Code, and there we see that it lacks in intellisense for jsdocs compared with IntelliJ/WebStorm, especially in these cases. Even WebStorm doesn't always correctly show everything, depending on the number of extends etc.... 

What we do is first play with the docblocks to get it correctly working in WebStorm, and then the VSCode guys pray it also works there. Mongoose Models for example are a very hard thing to document, but we found the 'perfect' solution for it after a lot of trial and error.....

**Comment by nicoespeon** [comment]
Hi there 👋 

I ran into the same issue and the best workaround I've found is [to use an intersection type](https://github.com/Microsoft/TypeScript/issues/20077#issuecomment-392712599).
 
```js
/**
 * @typedef {Object} ChildType
 * @property {String} childProp
 *
 * @typedef {Base & ChildType} Child
 */
```

**Comment by nemoDreamer** [comment]
Intersection type works great in most cases, but what about when your base type defines a required prop, but your child type makes it optional?

```
/**
 * @typedef {Object} Base
 * @prop {string} activeId
 */

/**
 * @typedef {Object} Extra
 * @prop {string} [activeId]
 */

// this will still give "activeId missing" errors in `Composed` component:
/** @type {Base & Extra} Composed */
```

I've also tried this:

```
/**
 * @typedef {Base} Composed
 * @prop {string} [activeId]
 */
```

**Comment by dietergeerts** [comment]
@nemoDreamer , from a functional view, what you are doing just makes no sense. `Composed` is both `Base` and `Extra`, where one requires `activeId` to be there, so that means it is required....

For extending typedefs, I know do the following, which works better in WebStorm:
```
/**
 * @typedef {IBase,IComposed} Composed
 */

/**
 * @typedef {IComposed}
 * @property {string} activeId
 */
```

If extending a typedef like the first block, you can't add props to the same block, otherwise it doesn't work. In some cases, this still isn't completely fail save, but it's the best we found up until now.

**Comment by rbiggs** [comment]
@nemoDreamer, I believe that with type intersections where one object has a required type that another has it optional, the former wins out. The only way around this would be to define another type without the required property and another type with optional properties then combining them:

```javascript
/**
 * @typedef {Object} Person
 * @prop {string} name
 * @prop {number} age
 */
/**
 * @typedef {Object} Job
 * @prop {string} [job]
 */

/**
 * @typedef {Person & Job} Bozo
 */

/**
 * @type {Bozo}
 */
const bozo1 = {
  name: 'Bozo the Clown',
  age: 50,
  job: 'clown'
}
// or:
const bozo2 = {
  name: 'Bozo the Clown',
  age: 50
}
```

**Comment by GioLogist** [comment]
Not sure if [this](https://stackoverflow.com/questions/42879269/vscode-not-showing-proper-params-in-object) is related, but intellisense doesn't seem to be showing an Object typedef as shown below.

![Screenshot 2019-11-28 10 30 23](https://user-images.githubusercontent.com/4120754/69827143-f2098e80-11ca-11ea-8702-7d6babc9b360.png)

In this screenshiot, I'd expect `SearchOptions` to show the properties.

**Comment by rbiggs** [comment]
@GioLogist, that looks right to me. I would expect it to show anything other that what the type is for that parameter, in this case, `SearchOptions`. Are you in VSCode or what editor? If you're using VSCode then you can add a configuration file to turn on live type checking, which will give you even more type safety. For VSCode, in your project's root create a folder named `.vscode` (notice the leading dot, it's supposed to be invisible). Inside that create a file `settings.json`. In that put the following:
```javascript
{
  "javascript.implicitProjectConfig.checkJs": true
}
```
<img width="1010" alt="Screen Shot 2019-11-29 at 4 33 22 AM" src="https://user-images.githubusercontent.com/249218/69869474-cab1d080-1261-11ea-91fe-0af1d36dfb56.png">

And here's what happen's with your code and this setting. Notice how `getSearchTerms` has red squiggle underlining, indicating something is wrong. And hovering you get the reason, missing parameters:

<img width="1010" alt="Screen Shot 2019-11-29 at 4 34 15 AM" src="https://user-images.githubusercontent.com/249218/69869545-03ea4080-1262-11ea-9e12-0dd17b2aef61.png">

If we provide the arguments as defined in the JSDoc types, everybody is happy and on hover we can see what the type of the argument is:

<img width="1010" alt="Screen Shot 2019-11-29 at 4 34 54 AM" src="https://user-images.githubusercontent.com/249218/69869629-31cf8500-1262-11ea-8ff5-99c6191978b0.png">

However, if we screw up and use the wrong type in our parameters, such as giving `where` a number instead of a string, we are informed about the error:

<img width="1010" alt="Screen Shot 2019-11-29 at 4 35 43 AM" src="https://user-images.githubusercontent.com/249218/69869654-4a3f9f80-1262-11ea-8338-9f97ce8a0833.png">

**Comment by GioLogist** [comment]
@rbiggs Interesting! Thanks for the insightful follow up! So I guess my question here is, there's no setting that can be enabled that will show the expected properties for that object? 

In the [original](https://stackoverflow.com/questions/42879269/vscode-not-showing-proper-params-in-object) SO post I found, OP suggested he expected something along the lines of this screenshot. Which is what I too would expect as well.

![image](https://user-images.githubusercontent.com/4120754/69891523-1724fc80-12b2-11ea-9b50-7268f1211ad6.png)

**Comment by rbiggs** [comment]
I doubt it. You see, Intellisense is trying to show a call signature for the function. It's a common thing for intellisense to do. It just shows the types for the call signature. What if one of the parameters was an object type that had two dozen properties? That gets really hard to show as well as to grok. 

Now if you wanted to organize your code a bit. You could have the `getSearchTerms` in one file with its type definition. Then import it and its type definition into another file where you use it. Here the source file:

<img width="1010" alt="Screen Shot 2019-11-29 at 2 52 52 PM" src="https://user-images.githubusercontent.com/249218/69892262-8271cd00-12b8-11ea-8635-3164d03d4602.png">

Then the file we import it into:

<img width="1010" alt="Screen Shot 2019-11-29 at 2 53 38 PM" src="https://user-images.githubusercontent.com/249218/69892264-8dc4f880-12b8-11ea-9222-93094080983d.png">

In the file where we import the function, we can right click on it and choose `peek definition`:

<img width="1010" alt="Screen Shot 2019-11-29 at 2 54 14 PM" src="https://user-images.githubusercontent.com/249218/69892273-b0571180-12b8-11ea-9b00-baf0e65391c9.png">

This will give us the following popup, where we can see the original source definition for this function, as well as its type definition:

<img width="1010" alt="Screen Shot 2019-11-29 at 2 54 38 PM" src="https://user-images.githubusercontent.com/249218/69892286-d11f6700-12b8-11ea-8bbb-5ae315add03e.png">

We can also hover over the imported type and see the type structure, which is more what you were looking for. This always works like this for imported types:

<img width="1010" alt="Screen Shot 2019-11-29 at 2 54 58 PM" src="https://user-images.githubusercontent.com/249218/69892298-ee543580-12b8-11ea-887f-60cb59ea85b8.png">

**Comment by GioLogist** [comment]
> What if one of the parameters was an object type that had two dozen properties? That gets really hard to show as well as to grok.

Good point.

> Now if you wanted to organize your code a bit. You could have the getSearchTerms in one file with its type definition. Then import it and its type definition into another file where you use it. Here the source file:

In the context of a library, I don't think the general user would import type definitions in order to get this benefit. Although, I can see the value for local dev. And more than anything, I learned something new today 🤓 Thanks @rbiggs !

**Comment by rbiggs** [comment]
Yup, but `peek definition` will show  you a popup of any function/object that was imported from a module, and `go to definition` will take you directly to the definition in the source module. Besides, showing only type information on hover, while useful, is secondary to really good descriptions. TypeScript's Intellisense is so good because it provides a lot more than just type information. It also provides detail descriptions based on MDN documentation. You can spice things up by using the `@exmple` tag to show how to use the code. 

Here's an example of that in an NPM module that I published:

<img width="711" alt="Screen Shot 2019-11-30 at 2 56 33 AM" src="https://user-images.githubusercontent.com/249218/69899579-3efd7a00-131d-11ea-9052-e3d72e9cee85.png">

And scrolling down you can see the full `@example` and the JSDoc types at the end:

<img width="711" alt="Screen Shot 2019-11-30 at 2 56 49 AM" src="https://user-images.githubusercontent.com/249218/69899588-5a688500-131d-11ea-92db-8aa9e7fc8c56.png">

And here's an example of a complex type consisting of a main type and a subtype. State has an items property which is `Item[]`. On hover we can see what `Item` type consists of. But if we try to use it we discover that it has only two properties `key: number` and `value: string`. Trying to use any other property will give us a warning. So through Intellisense and good JSDoc types everything does get exposed to the end user through Intellisense or discoverability.

<img width="711" alt="Screen Shot 2019-11-30 at 3 04 48 AM" src="https://user-images.githubusercontent.com/249218/69899649-4d986100-131e-11ea-968d-346926aab930.png">

<img width="711" alt="Screen Shot 2019-11-30 at 3 05 04 AM" src="https://user-images.githubusercontent.com/249218/69899650-51c47e80-131e-11ea-90a0-29330966db70.png">

<img width="711" alt="Screen Shot 2019-11-30 at 3 06 49 AM" src="https://user-images.githubusercontent.com/249218/69899660-7f112c80-131e-11ea-9164-59e48241e44b.png">

**Comment by rbiggs** [comment]
By the way, these same issues with types in Intellisense with JSDoc comments also apply to TypeScript types and Intellisense. It's the way Intellisense works. Now a way to get around this would be to enable the ability to drill down into types from the Intellisense popup, maybe by right clicking on a type in the popup. But that would probably introduce a level of complexity to the current Intellisense implementation that the VSCode team would not want to consider it. Kind of like introducing type inception, down a rabbit hole.

**Comment by GioLogist** [comment]
*Extremely* useful stuff here @rbiggs ! I see you're in Reno, happy to get you a beer next time i'm in Tahoe (pretty often), if you're up for it 🤘🏻

> Now a way to get around this would be to enable the ability to drill down into types from the Intellisense popup, maybe by right clicking on a type in the popup.

Really like this idea. Funny enough, I tried ctr+clicking the type from the popup originally, which is what led me here.

**Comment by rbiggs** [comment]
Woohoo! Join us at BeerJS in Reno: https://github.com/beerjs/reno. This December is quick talks. I'll be doing a whirlwind intro to the Elm Architecture for React. You might already know Jordan 
Papaleo, who used to run the JavaScript Meetup in Tahoe. ;-)

Oh yeah, control clicking stuff in VSCode can take you to unexpected places, he he.

**Comment by rbiggs** [comment]
Woohoo! Join us at BeerJS in Reno: https://github.com/beerjs/reno. This December is quick talks. I'll be doing a whirlwind intro to the Elm Architecture for React. You might already know Jordan
Papaleo, who used to run the JavaScript Meetup in Tahoe. ;-)

Oh yeah, control clicking stuff in VSCode can take you to unexpected places, he he.

**Comment by laruiss** [comment]
> 
> 
> @nicoespeon Which version of jsdoc are you using? I get
> 
> ```
> ERROR: Unable to parse a tag's type expression for source file 
> Invalid type expression "testBase & testSubType": Expected "|" but "&" found.
> ```

Take a look at this plugin:
https://github.com/chriseaton/jsdoc-plugin-intersection

**Comment by rbiggs** [comment]
Not sure what you'e referring to as a problem. Can you provide some code?

**Comment by laruiss** [comment]
Sorry, let me rephrase the whole thing:

@nicoespeon provided (half) a solution with the [intersection type](https://github.com/Microsoft/TypeScript/issues/20077#issuecomment-392712599) and this example snippet:

```
/**
 * @typedef {Object} ChildType
 * @property {String} childProp
 *
 * @typedef {Base & ChildType} Child
 */
```

The problem is that this syntax is not (and most likely will never be) recognized by `jsdoc`, hence the error mentioned by @henhal (when he runs the generation):

```
ERROR: Unable to parse a tag's type expression for source file 
Invalid type expression "testBase & testSubType": Expected "|" but "&" found.
```

But @chriseaton made this [plugin](https://github.com/chriseaton/jsdoc-plugin-intersection) and it works like a charm (I tested it) : it « Converts TypeScript intersection types (joined with an "&") to a jsDoc type union "|" allowing the file to be processed downstream. This allows you to use the amperstand "&" in your code. »

So when using this plugin, one can use the intersection syntax so that one has the right completion in VSCode (for example) and still has the ability to generate the doc as usual.

**Comment by dietergeerts** [comment]
> ```
> /**
>  * @typedef {Object} ChildType
>  * @property {String} childProp
>  *
>  * @typedef {Base & ChildType} Child
>  */

Can you try the following (which does work for our use-case very well, even our IDE can work with it:
```
/**
 * @typedef {Object} ChildType
 * @property {string} childProp
 */
/**
 * @typedef {Base, ChildType} Child
 */
```
Notice that they need to be in their own separate docblock to have this working.

**Comment by laruiss** [comment]
Unfortunately, my editor (VS Code) only takes the first type def:

![combined-jsdoc](https://user-images.githubusercontent.com/2486424/83144549-cc1a6380-a0f3-11ea-840e-2952d25e857a.gif)

**Comment by thw0rted** [comment]
Trackback to #1285 -- current stance is that intersection types won't be supported but I asked them to reconsider.  Maybe this is a helpful use-case in support of that?

**Comment by dietergeerts** [comment]
It works in WebStorm:

![image](https://user-images.githubusercontent.com/7393706/84664068-47ff1300-af1e-11ea-8a63-60eda0537899.png)

![image](https://user-images.githubusercontent.com/7393706/84664135-61a05a80-af1e-11ea-9d5e-0fb939fb17c8.png)

**Comment by dietergeerts** [comment]
And this is how I get it works in VSCode:

![image](https://user-images.githubusercontent.com/7393706/84664386-c2c82e00-af1e-11ea-8947-487c847024aa.png)

![image](https://user-images.githubusercontent.com/7393706/84664429-d4113a80-af1e-11ea-9262-c73abac896ed.png)

**Comment by dietergeerts** [comment]
And using the `|` also works in WebStorm apparently:

![image](https://user-images.githubusercontent.com/7393706/84664529-f5722680-af1e-11ea-9820-b2b8eff9e60a.png)

So the `|` symbol appears to be the way to go to have it working in both WebStorm and VSCode.

**Comment by thw0rted** [comment]
Those are two different things -- they both do what it says on the tin.  *Intersection* type (`A & B`) is a type containing the properties common to both A and B; *union* type (A | B) is all the properties in A plus all the properties in B.

**Comment by dietergeerts** [comment]
> Those are two different things -- they both do what it says on the tin. _Intersection_ type (`A & B`) is a type containing the properties common to both A and B; _union_ type (A | B) is all the properties in A plus all the properties in B.

I know they are different, but your explanation of both are not what the TS docs, nor the JSDoc docs are saying:

- https://www.typescriptlang.org/docs/handbook/advanced-types.html#intersection-types.
- https://jsdoc.app/tags-type.html

So it is:

- `&` means it will both be A and B, so all props of A and all props of B (This only exists in TS)
- `|` means it will either be A or B, so OR it is A, OR it is B, but not both (This exists in both TS and JSDoc)

**Comment by thw0rted** [comment]
You're right, I didn't explain it well, probably because I haven't thought it all the way through myself.

With Intersection, it is known to be an A "and" a B, at the same time, while with a Union it's known to be A "or" B but you can't tell which without further checking.  My mental model of a Union is that you can access properties that are common to A and B, or you can make a runtime type check to narrow down which one you actually have.  That's why I say it has "all the properties", because it *could* have them at runtime.

When I say that the Intersection contains "properties common to both" I was specifically thinking of the way that Intersections can make possible types that otherwise can't exist.  I came to this issue because I currently use `type DictionaryClass = {a(): string, b(): number} & {[k: string]: boolean}` to describe a type whose `a` and `b` properties are functions but where any other property is a boolean.  Normally the index signature would make declaring `a` or `b` impossible, because they're not booleans, and that's why the merging behavior is so important.

**Comment by dietergeerts** [comment]
@thw0rted , no prob. Thx for that example.

**Comment by yunnysunny** [comment]
I found the best solution on here: https://stackoverflow.com/questions/36737921/how-to-extend-a-typedef-parameter-in-jsdoc
See the second answer, it suitable for jsdoc.

**Comment by AnrDaemon** [comment]
What I'd like to know is a very simple use case:

```js
const utils = {
    "meta_chars": new RegExp('[\\^$.|?*+()[{]', 'g'),
    /** Escapes regexp meta characters
     *
     * @param {string} string A string to be quoted.
     * @returns {string} Quoted string.
     */
    "preg_quote": (string) => string.replace(utils.meta_chars, '\\$&'),
}

/** ???
 */
String.prototype.quote = (string) => utils.preg_quote(string);
```

How should I document this thing to see the hint in other places of the project?

**Comment by TicTak21** [comment]
Apparently still doesn't work.

**Comment by rikka0w0** [comment]
``` Javascript
/**
 * Base
 * 
 * @typedef {Object} Base
 * @property {string} prop1
 * @property {string} prop2
 */

/**
 * Represents a local outbound.
 * @typedef {Base& {
 * 	prop1: 'value_case_1', // Override something in the Base type
 * 	prop2: string, // Unique property in the Child type
 * }} Child
 */
```

This works in VSCode.

**Comment by RiQuY** [comment]
**8 years**, still no official JSDoc solution.


![](https://c.tenor.com/tDpElw3fCiQAAAAC/ez-clap.gif)
